api/v1alpha1/openstackimage_types.go

```diff
@@ -0,0 +1,218 @@
+}
+
+// +kubebuilder:validation:Enum:=xz;gz
+type ImageCompression string
+
+const (
+	ImageCompressionXZ ImageCompression = "xz"
+	ImageCompressionGZ ImageCompression = "gz"
+)
+
+type ImageContent struct {
+	// Hash is a hash which can be used to verify the downloaded image data
+	Hash *ImageHash `json:"hash,omitempty"`
+
+	// GlanceHashAlgorithm is the algorithm of the hash glance will publish.
+	// If not set it will default to sha512, which is Glance's default. It
+	// MUST be set to the value configured in Glance.
+	// +kubebuilder:default:=sha512
+	// +required
+	GlanceHashAlgorithm *ImageHashAlgorithm `json:"glanceHashAlgorithm,omitempty"`
+
+	// Source specifies how to obtain the image data
+	Source ImageSource `json:"source"`
+}
+
+type ImageSourceType string
+
+const (
+	ImageSourceTypeURL ImageSourceType = "url"
+)
+
+// ImageSource specifies the source of image data
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'url' ?  has(self.url) : !has(self.url)",message="url is required when type is url, and forbidden otherwise"
+// +union
+type ImageSource struct {
+	// Type is the type of the image source
+	// +kubebuilder:validation:Required
+	// +unionDiscriminator
+	Type ImageSourceType `json:"type"`
+
+	// URL describes how to obtain image data by downloading it from a URL. Must be set if Type is 'url'
+	// +unionMember
+	URL *ImageSourceURL `json:"url,omitempty"`
+}
+
+type ImageSourceURL struct {
+	// URL containing image data
+	// +kubebuilder:validation:Format=uri
+	// +kubebuilder:validation:Required
+	URL string `json:"url"`
+
+	// Decompress specifies that the source data must be decompressed with the given compression algorithm before being stored
+	// +optional
+	Decompress *ImageCompression `json:"decompress,omitempty"`
+}
+
+type ImageHash struct {
+	// Algorithm is the hash algorithm used to generate value.
+	// +required
+	Algorithm ImageHashAlgorithm `json:"algorithm"`
+
+	// Value is the hash of the image data using Algorithm.
+	// kubebuilder:validation:MinLength:=1
+	// +required
+	Value string `json:"value"`
+}
+
+// OpenStackImageSpec defines the desired state of OpenStackImage.
+type OpenStackImageSpec struct {
+	// ContainerFormat is the format of the image container.
+	// qcow2 and raw images do not usually have a container, and this can be omitted.
+	// +optional
+	ContainerFormat *ImageContainerFormat `json:"containerFormat,omitempty"`
+
+	// DiskFormat is the format of the disk image.
+	// Normal values are "qcow2", or "raw". Glance may be configured to support others.
+	// +optional
+	DiskFormat *ImageDiskFormat `json:"diskFormat,omitempty"`
+
+	// Protected specifies that the image is protected from deletion.
+	// If not specified, the default is false.
+	// +optional
+	Protected *bool `json:"protected,omitempty"`
+
+	// Tags is a list of tags which will be applied to the image. A tag has a maximum length of 255 characters.
+	// +optional
+	Tags []ImageTag `json:"tags,omitempty"`
```

Comment:
Need an SSA `listType` tag, is this a map list with a key? Else mark explicitly as atomic